// amadeus/lib/amadeus/client/errors.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
 * The error that is passed to the Promise when the API call fails.
 *
 * @param {Response} response the {@link Response} object containing the raw
 *  http response and the {@link Request} instance that made the API call.
 * @property {Response} response the {@link Response} object containing the raw
 *  http response and the {@link Request} instance that made the API call.
 * @property {string} code a unique code for this type of error. Options include
 *  `NetworkError`, `ParserError`, `ResponseError`, `ServerError`,
 *  `AuthenticationError`, `NotFoundError` and `UnknownError`
 *  from the  {@link Response}'s parsed data
 */
var ResponseError = exports.ResponseError = function () {
  function ResponseError(response) {
    _classCallCheck(this, ResponseError);

    this.response = response;
    this.determineDescription();
  }

  // PRIVATE

  _createClass(ResponseError, [{
    key: 'determineDescription',
    value: function determineDescription() {
      if (!this.response || !this.response.parsed) {
        this.description = null;
        return;
      }
      var result = this.response.result;
      if (result && result.errors) {
        this.description = result.errors;
      } else if (result) {
        this.description = result;
      }
      return;
    }
  }]);

  return ResponseError;
}();

// Protected

var NetworkError = exports.NetworkError = function (_ResponseError) {
  _inherits(NetworkError, _ResponseError);

  function NetworkError() {
    var _ref;

    _classCallCheck(this, NetworkError);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    var _this = _possibleConstructorReturn(this, (_ref = NetworkError.__proto__ || Object.getPrototypeOf(NetworkError)).call.apply(_ref, [this].concat(args)));

    _this.code = 'NetworkError';
    return _this;
  }

  return NetworkError;
}(ResponseError);

var ParserError = exports.ParserError = function (_ResponseError2) {
  _inherits(ParserError, _ResponseError2);

  function ParserError() {
    var _ref2;

    _classCallCheck(this, ParserError);

    for (var _len2 = arguments.length, args = Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
      args[_key2] = arguments[_key2];
    }

    var _this2 = _possibleConstructorReturn(this, (_ref2 = ParserError.__proto__ || Object.getPrototypeOf(ParserError)).call.apply(_ref2, [this].concat(args)));

    _this2.code = 'ParserError';
    return _this2;
  }

  return ParserError;
}(ResponseError);

var ServerError = exports.ServerError = function (_ResponseError3) {
  _inherits(ServerError, _ResponseError3);

  function ServerError() {
    var _ref3;

    _classCallCheck(this, ServerError);

    for (var _len3 = arguments.length, args = Array(_len3), _key3 = 0; _key3 < _len3; _key3++) {
      args[_key3] = arguments[_key3];
    }

    var _this3 = _possibleConstructorReturn(this, (_ref3 = ServerError.__proto__ || Object.getPrototypeOf(ServerError)).call.apply(_ref3, [this].concat(args)));

    _this3.code = 'ServerError';
    return _this3;
  }

  return ServerError;
}(ResponseError);

var ClientError = exports.ClientError = function (_ResponseError4) {
  _inherits(ClientError, _ResponseError4);

  function ClientError() {
    var _ref4;

    _classCallCheck(this, ClientError);

    for (var _len4 = arguments.length, args = Array(_len4), _key4 = 0; _key4 < _len4; _key4++) {
      args[_key4] = arguments[_key4];
    }

    var _this4 = _possibleConstructorReturn(this, (_ref4 = ClientError.__proto__ || Object.getPrototypeOf(ClientError)).call.apply(_ref4, [this].concat(args)));

    _this4.code = 'ClientError';
    return _this4;
  }

  return ClientError;
}(ResponseError);

var AuthenticationError = exports.AuthenticationError = function (_ResponseError5) {
  _inherits(AuthenticationError, _ResponseError5);

  function AuthenticationError() {
    var _ref5;

    _classCallCheck(this, AuthenticationError);

    for (var _len5 = arguments.length, args = Array(_len5), _key5 = 0; _key5 < _len5; _key5++) {
      args[_key5] = arguments[_key5];
    }

    var _this5 = _possibleConstructorReturn(this, (_ref5 = AuthenticationError.__proto__ || Object.getPrototypeOf(AuthenticationError)).call.apply(_ref5, [this].concat(args)));

    _this5.code = 'AuthenticationError';
    return _this5;
  }

  return AuthenticationError;
}(ResponseError);

var NotFoundError = exports.NotFoundError = function (_ResponseError6) {
  _inherits(NotFoundError, _ResponseError6);

  function NotFoundError() {
    var _ref6;

    _classCallCheck(this, NotFoundError);

    for (var _len6 = arguments.length, args = Array(_len6), _key6 = 0; _key6 < _len6; _key6++) {
      args[_key6] = arguments[_key6];
    }

    var _this6 = _possibleConstructorReturn(this, (_ref6 = NotFoundError.__proto__ || Object.getPrototypeOf(NotFoundError)).call.apply(_ref6, [this].concat(args)));

    _this6.code = 'NotFoundError';
    return _this6;
  }

  return NotFoundError;
}(ResponseError);

var UnknownError = exports.UnknownError = function (_ResponseError7) {
  _inherits(UnknownError, _ResponseError7);

  function UnknownError() {
    var _ref7;

    _classCallCheck(this, UnknownError);

    for (var _len7 = arguments.length, args = Array(_len7), _key7 = 0; _key7 < _len7; _key7++) {
      args[_key7] = arguments[_key7];
    }

    var _this7 = _possibleConstructorReturn(this, (_ref7 = UnknownError.__proto__ || Object.getPrototypeOf(UnknownError)).call.apply(_ref7, [this].concat(args)));

    _this7.code = 'UnknownError';
    return _this7;
  }

  return UnknownError;
}(ResponseError);
