// amadeus/lib/amadeus/namespaces/reference_data.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _urls = require('./reference_data/urls');

var _urls2 = _interopRequireDefault(_urls);

var _locations = require('./reference_data/locations');

var _locations2 = _interopRequireDefault(_locations);

var _location = require('./reference_data/location');

var _location2 = _interopRequireDefault(_location);

var _airlines = require('./reference_data/airlines');

var _airlines2 = _interopRequireDefault(_airlines);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
 * A namespaced client for the
 * `/v2/reference-data` endpoints
 *
 * Access via the {Amadeus} object
 *
 * ```js
 * let amadeus = new Amadeus();
 * amadeus.referenceData.urls;
 * ```
 *
 * @param {Client} client
 * @property {Urls} urls
 * @protected
 */
var ReferenceData = function () {
  function ReferenceData(client) {
    _classCallCheck(this, ReferenceData);

    this.client = client;
    this.urls = new _urls2.default(client);
    this.locations = new _locations2.default(client);
    this.airlines = new _airlines2.default(client);
  }

  /**
   * The namespace for the Location APIs - accessing a specific location
   *
   * @param  {string} [locationId]  The ID of the location to search for
   * @return {Location}
   **/


  _createClass(ReferenceData, [{
    key: 'location',
    value: function location(locationId) {
      return new _location2.default(this.client, locationId);
    }
  }]);

  return ReferenceData;
}();

exports.default = ReferenceData;
module.exports = exports['default'];
