// amadeus/lib/amadeus/namespaces/shopping.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _flight_destinations = require('./shopping/flight_destinations');

var _flight_destinations2 = _interopRequireDefault(_flight_destinations);

var _flight_offers = require('./shopping/flight_offers');

var _flight_offers2 = _interopRequireDefault(_flight_offers);

var _flight_dates = require('./shopping/flight_dates');

var _flight_dates2 = _interopRequireDefault(_flight_dates);

var _hotel_offers = require('./shopping/hotel_offers');

var _hotel_offers2 = _interopRequireDefault(_hotel_offers);

var _hotel_offers_by_hotel = require('./shopping/hotel_offers_by_hotel');

var _hotel_offers_by_hotel2 = _interopRequireDefault(_hotel_offers_by_hotel);

var _hotel_offer = require('./shopping/hotel_offer');

var _hotel_offer2 = _interopRequireDefault(_hotel_offer);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
 * A namespaced client for the
 * `/v1/shopping` endpoints
 *
 * Access via the {@link Amadeus} object
 *
 * ```js
 * let amadeus = new Amadeus();
 * amadeus.shopping;
 * ```
 *
 * @param {Client} client
 * @property {FlightDestinations} flight_destinations
 * @property {FlightOffers} flight_offers
 * @property {FlightDates} flight_dates
 * @property {HotelOffers} hotel_offers
 * @property {HotelOffer} hotel_offer
 * @property {HotelOffersByHotel} hotel_offers_by_hotel
 */
var Shopping = function () {
  function Shopping(client) {
    _classCallCheck(this, Shopping);

    this.client = client;
    this.flightDestinations = new _flight_destinations2.default(client);
    this.flightOffers = new _flight_offers2.default(client);
    this.flightDates = new _flight_dates2.default(client);
    this.hotelOffers = new _hotel_offers2.default(client);
    this.hotelOffersByHotel = new _hotel_offers_by_hotel2.default(client);
  }

  /**
   * Loads a namespaced path for a specific offer ID
   *
   * @param  {string} [offerId]  The ID of the offer for a dedicated hotel
   * @return {HotelOffer}
   **/


  _createClass(Shopping, [{
    key: 'hotelOffer',
    value: function hotelOffer(offerId) {
      return new _hotel_offer2.default(this.client, offerId);
    }
  }]);

  return Shopping;
}();

exports.default = Shopping;
module.exports = exports['default'];
