// amadeus/lib/amadeus/namespaces/travel/analytics/air_traffic.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _traveled = require('./air_traffic/traveled.js');

var _traveled2 = _interopRequireDefault(_traveled);

var _booked = require('./air_traffic/booked.js');

var _booked2 = _interopRequireDefault(_booked);

var _searched = require('./air_traffic/searched.js');

var _searched2 = _interopRequireDefault(_searched);

var _searched_by_destination = require('./air_traffic/searched_by_destination.js');

var _searched_by_destination2 = _interopRequireDefault(_searched_by_destination);

var _busiest_period = require('./air_traffic/busiest_period.js');

var _busiest_period2 = _interopRequireDefault(_busiest_period);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
 * A namespaced client for the
 * `/v1/travel/analytics/air-traffic` endpoints
 *
 * Access via the {@link Amadeus} object
 *
 * ```js
 * let amadeus = new Amadeus();
 * amadeus.travel.analytics.airTraffic;
 * ```
 *
 * @param {Client} client
 */
var AirTraffic = function AirTraffic(client) {
  _classCallCheck(this, AirTraffic);

  this.client = client;
  this.traveled = new _traveled2.default(client);
  this.booked = new _booked2.default(client);
  this.searched = new _searched2.default(client);
  this.searchedByDestination = new _searched_by_destination2.default(client);
  this.busiestPeriod = new _busiest_period2.default(client);
};

exports.default = AirTraffic;
module.exports = exports['default'];
